### PR TITLE
Auto fetch Browsers

### PR DESCRIPTION
**Pull request by jingt06**
- Collect historical browsers APIs by specifying command line argument historical (node main/selenium_og.es6.js --historical)
- Fetch all browsers information from https://www.browserstack.com/automate/browsers.json, and filtered by config file browsers_env.json
- Run with 50 different versions/platforms of Chrome, Firefox, Edge and Safari. All completed.

### CODE REVIEW

File: browsers_envs.json [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    "min_browser_version" : "13",
+    "os" : [
+      {
+        "os" : "Windows",
+        "os_version" : "10"
+      }
+    ],
+    "resolution" : "1024x768"
+  }
+]
```
Comment (mdittmer) [inline browsers_envs.json]: nit: Please configure your editor to include a newline at the end of files. Please use your editor or command-line tools to make sure files touched by this PR all end with `\n`.

File: data/og/window_Firefox_48.0_OSX_10.6.json [hunk 2]
```diff
@@ -0,0 +1,2 @@
+{
+}
```
Comment (mdittmer) [inline data/og/window_Firefox_48.0_OSX_10.6.json]: Did this data collection fail? The database is empty. If we can't collect data from this version yet, let's not check in an empty database (and let's also file a bug).

File: data/og/window_IE_11.0_Windows_6.3.json [hunk 3]
```diff
@@ -0,0 +1,2 @@
+{
+}
```
Comment (mdittmer) [inline data/og/window_IE_11.0_Windows_6.3.json]: ditto: empty database.

File: lib/get_browsers.es6.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+ */
+'use strict';
+
+const fs = require('fs');
+const https = require('https');
+const env = require('process').env;
+const argv = require('yargs').argv;
+
+module.exports = new Promise(function(resolve, reject) {
+  // if historical is not defined as commandline args,
```
Comment (mdittmer) [inline lib/get_browsers.es6.js]: nit: Comment-as-sentence; please add capital and period.

File: lib/get_browsers.es6.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  // fetch list of browsers info from envs file
+  if ( !argv.historical ) {
+    const browsers = JSON.parse(
+      fs.readFileSync(`${__dirname}/../${env.SELENIUM_HOST}_envs.json`)
```
Comment (mdittmer) [inline lib/get_browsers.es6.js]: nit: Please add semi-colon. I recommend setting up ESLint to alert you of missing semi-colons in your editor.

File: lib/get_browsers.es6.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    );
+    resolve( browsers );
+    return
```
Comment (mdittmer) [inline lib/get_browsers.es6.js]: ditto: missing semi-colon.

File: lib/get_browsers.es6.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  // if historical is defined as command line args,
```
Comment (mdittmer) [inline lib/get_browsers.es6.js]: nit: Comment-as-sentence.

File: lib/get_browsers.es6.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  // fetch all supported browsers from
```
Comment (mdittmer) [inline lib/get_browsers.es6.js]: Please update this comment to reflect that we will use the subset of `www.browserstack.com/automate/browsers.json`'s browsers that match `browser_envs.json`.

File: lib/get_browsers.es6.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+	}, function(res) {
+	  var body = '';
+
+	  res.on('data', function(chunk){
+	      body += chunk;
+	  });
+
+	  res.on('end', function(){
+	    allBrowsers = JSON.parse(body);
+      	// filter allBrowsers by config from browsers_env.json
```
Comment (mdittmer) [inline lib/get_browsers.es6.js]: nits: Indent to same level as code; comment-as-sentence.

File: lib/get_browsers.es6.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    );
+    resolve( browsers );
+    return
+  }
+
+  // if historical is defined as command line args,
+  // fetch all supported browsers from
+  // www.browserstack.com/automate/browsers.json
+  var browsers = [];
+	var allBrowsers = null;
```
Comment (mdittmer) [inline lib/get_browsers.es6.js]: Why does indentation jump here? It shouldn't.

File: lib/get_browsers.es6.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+	            browserFilter.browserName.toLowerCase() &&
+	          parseFloat( browser.browser_version ) >=  // browser_version
+	            parseFloat( browserFilter.min_browser_version ) &&
+	          browserFilter.os.filter( os => {          // os and os_version
+	            return os.os === browser.os &&
+	            os.os_version === browser.os_version;
+	          }).length !== 0
+	      }));
+	    });
+      browsers = browsers.map( browser => {           // add browserName field
```
Comment (mdittmer) [inline lib/get_browsers.es6.js]: nits: no space between `map(` and `browser`; just one space between `{` and `//`; comment-as-sentence.

File: lib/get_browsers.es6.js [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        browser.browserName = browser.browser;
+        return browser;
+      });
+      resolve( browsers );
```
Comment (mdittmer) [inline lib/get_browsers.es6.js]: nit: no spaces around `browsers`.

File: main/selenium_og.es6.js [hunk 13]
```diff
@@ -22,163 +22,169 @@
 const fs = require('fs');
 const webdriver = require('selenium-webdriver');
 const hostModule = require(`../lib/remote/selenium/selenium-host.js`);
-const browsers = JSON.parse(
-  fs.readFileSync(`${__dirname}/../${hostModule.name}_envs.json`)
-);
+
```
Comment (mdittmer) [inline main/selenium_og.es6.js]: nit: just one empty line for spacing.

File: main/selenium_og.es6.js [hunk 14]
```diff
@@ -22,163 +22,169 @@
...
 const loggerModule = require('../lib/logger.es6.js');
 const throttle = require('../lib/throttle.es6.js');
 const NameRewriter = require('object-graph-js').NameRewriter;
 const stringify = require('ya-stdlib-js').stringify;
+const getBrowsers = require('../lib/get_browsers.es6.js');
+
 
 const By = webdriver.By;
+getBrowsers.then( browsers => {
```
Comment (mdittmer) [inline main/selenium_og.es6.js]: nit: no space between `then(` and `browsers`.

**Comment by jingt06** [comment]
- Used ESlint to lint the code.
- Run test on IE (version>10) on windows 7, 8 , 10.

File: lib/get_browsers.es6.js (hunk 8 above)
Comment (mdittmer) [inline lib/get_browsers.es6.js]: @jingt06 I don't think this comment has been addressed yet.

Comment (jingt06) [inline lib/get_browsers.es6.js]: @mdittmer , updated

**Pull request merged**